### ISSUE TITLE
TypeScript is OK with an arrow function returning more than an explicit void, but complains when a declared function is doing the same.

### ISSUE THREAD

**Issue by Protectator** [microsoft/TypeScript#45247]
# Bug Report

TypeScript is OK with an arrow function returning more than an explicit void, but complains when a declared function is doing the same.

### 🔎 Search Terms

void, function, arrow function

### 🕗 Version & Regression Information

I can't find any prior version where this behavior is explainable.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAZgRgFwwBQEoYF4B8MBuIAlgCbZqa4wAsA3DAPQMwA8AtDAPIDSAsAFAC4AVzDAoRcPABMGFIVIwA3gJgwATgFMow9WBr0mrDgBUAngAdNMAORhhAWwBGm9TZhEIMMCFgBDCAgiAHMwPycAG2soEBgoS2sbBRIbADoBAF8gA)

### 💻 Code

```ts
const f1: () => void = () => 4; // <- OK

function f2(): void {
  return 4; // <- Type 'number' is not assignable to type 'void'.
}
```

### 🙁 Actual behavior

TypeScript complains that "Type 'number' is not assignable to type 'void'." in a function that returns void.

### 🙂 Expected behavior

TypeScript doesn't complains because `number` gives more information than `void`.

**Comment by MartinJohns** [comment]
The arrow function in `f1` doesn't have an explicit void. It has no explicit return type defined, so it's inferred.

When you actually set an explicit return type `void` you get the same error:

    const f1: () => void = (): void => 4; // <- Type 'number' is not assignable to type 'void'.

**Comment by captain-yossarian** [comment]
@Protectator please link SO question so @MartinJohns will be able to answer it

**Comment by Protectator** [comment]
Right, thanks for the answe. Here is it : https://stackoverflow.com/questions/68588910/why-is-typescript-ok-with-an-arrow-function-returning-more-than-an-explicit-void
